File: 2022/06/16_regular_expression_matching/yongsu.c
#include <stdio.h>
#include <stdbool.h>

// Runtime: 65 ms, faster than 16.32% of C online submissions for Regular Expression Matching.
// Memory Usage: 5.7 MB, less than 20.53% of C online submissions for Regular Expression Matching.

int my_strlen(char * s) {
    int len = 0;
    while (s[len]) len++;
    return len;
}

bool dfs(char * s, char * p) {
    bool s_is_done = (my_strlen(s) == 0); // s가 끝났는지 확인
    bool p_is_done = (my_strlen(p) == 0); // p가 끝났는지 확인
    
    if (p_is_done) return s_is_done; // p가 끝났는데 s도 끝났으면 성공, 아니면 실패

    if (p[1] == '*') { // *은 문자열이 반복되지 않거나 (0개) 반복될 경우(1개 이상) 둘 다 확인해야함
        // 0개라고 가정하고 패턴 2칸 건너뛰기, 얘는 s가 끝나도 p가 끝나면 통과기 때문에 s가 끝났는지 확인 안 해도됨
        return (dfs(s, p + 2)) || // 두 케이스 중 하나만 만족하면 되니까 OR 연산
        // s가 안끝났고, 똑같은 문자가 1개 이상이라고 가정하고 s를 한칸씩 움직이면서 체크해야하는데, 이 때 s첫글자랑 p첫글자가 같거나, p첫글자가 .이어야 함. 같다면 다음 문자 확인
        (!s_is_done && (s[0] == p[0] || p[0] == '.') && dfs(s + 1, p));
    } else {
        // s가 안끝났고, s와 p 첫글자가 같거나 p 첫글자가 .이면, 통과하고 그 다음 문자열 비교
        return !s_is_done && (s[0] == p[0] || p[0] == '.') && dfs(s + 1, p + 1);
    }
}

bool isMatch(char * s, char * p) {
    return dfs(s, p);
}

// int main(void) {
//     printf("%d\n", isMatch("aa", "a"));  //0
//     printf("%d\n", isMatch("aa", "a*")); //1
//     printf("%d\n", isMatch("aaaaaa", "a*")); //1
//     printf("%d\n", isMatch("aaa", "a*a")); //1
//     printf("%d\n", isMatch("aaaaaa", ".*")); //1
//     printf("%d\n", isMatch("ab", ".*")); //1
//     printf("%d\n", isMatch("aab", "c*a*b")); //1
//     return 0;
// }

File: 2022/06/21_letter_combination_of_a_phone_number/yongsu.py
from typing import List
import unittest

# Runtime: 64 ms, faster than 7.91% of Python3 online submissions for Letter Combinations of a Phone Number.
# Memory Usage: 13.9 MB, less than 29.76% of Python3 online submissions for Letter Combinations of a Phone Number.

# C로 짜기 연습하자 제발

class Solution:
    letters = {}

    def __init__(self) -> None:
        self.letters[2] = "abc"
        self.letters[3] = "def"
        self.letters[4] = "ghi"
        self.letters[5] = "jkl"
        self.letters[6] = "mno"
        self.letters[7] = "pqrs"
        self.letters[8] = "tuv"
        self.letters[9] = "wxyz"

    # 전부 다 순회하면서 하나씩 붙이기
    def letterCombinations(self, digits: str) -> List[str]:
        ans = []

        def dfs(index: int, buf: str, leftover: str):
            if len(buf) >= len(digits):
                ans.append(buf)
                return
            
            for i in range(len(leftover)):                  # 전체 숫자를 다 돌면서 할거임
                for j in self.letters[int(leftover[i])]:    # 그 숫자부터 시작해서 나머지 숫자들로 만들 수 있는 조합을 만들기
                    dfs(i + 1, buf + j, leftover[i + 1:])   # 현재 보고 있는 숫자 이전은 다시 조합할 필요 없으므로 내가 보려는 숫자는 그 숫자들 이전은 모두 제외한 숫자만 봐야함

        if len(digits) == 0:
            return []

        dfs(0, "", digits)

        return ans

class SolutionTests(unittest.TestCase):
    def setUp(self) -> None:
        self.sol = Solution()

    def test_1(self):
        answer = ["ad","ae","af","bd","be","bf","cd","ce","cf"]
        result = self.sol.letterCombinations("23")
        self.assertEqual(result, answer)


if __name__ == "__main__":
    unittest.main()